agent/processwatcher.go

```diff
@@ -0,0 +1,434 @@
+	for i, library := range pm.libraries {
+		if library.Path() == path {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func (pm *processMap) maybeAddLibrary(lib libspector.Library) int {
+	path := lib.Path()
+	index := pm.findLibraryIndex(path)
+
+	if index < 0 {
+		pm.libraries = append(pm.libraries, lib)
+		index = pm.findLibraryIndex(path)
+	}
+
+	return index
+}
+
+func (wt *processWatcher) acquireState() *processMap {
+	procs, err := wt.processes()
+	if err != nil {
+		panic(fmt.Errorf("Couldn't load processes: %s", err))
+	}
+
+	pm := processMap{
+		processes: make(systemProcesses, 0, len(procs)),
+		libraries: make(systemLibraries, 0), // ¯\_(ツ)_/¯
+	}
+
+	for _, proc := range procs {
+		started, err := proc.Started()
+		if err != nil {
+			log.Infof("PID %d is not running, skipping", proc.PID())
+			continue
+		}
+
+		command, err := proc.Command()
+		if err != nil {
+			log.Infof("Can't read command line for PID %d: %v", proc.PID(), err)
+			// fall through, we can live without this (?)
+		}
+
+		spectorLibs, err := proc.Libraries()
+		if err != nil {
+			if os.Getuid() != 0 && os.Geteuid() != 0 {
+				log.Infof("Cannot examine libs for PID %d, with UID:%d, EUID:%d",
+					proc.PID(), os.Getuid(), os.Geteuid())
+				continue
+			}
+
+			// otherwise barf
+			panic(fmt.Errorf("Couldn't load libs for process %v: %s", proc, err))
+		}
+
+		wp := watchedProcess{
+			ProcessStartedAt: started,
+			Pid:              proc.PID(),
+			Libraries:        make([]processLibrary, len(spectorLibs)),
+			Outdated:         false,
+			Command:          command,
+		}
+
+		for i, spectorLib := range spectorLibs {
+			libraryIndex := pm.maybeAddLibrary(spectorLib)
+			lib := processLibrary{
+				libraryIndex: libraryIndex,
+				Outdated:     spectorLib.Outdated(proc),
+			}
+			if lib.Outdated && !wp.Outdated {
+				wp.Outdated = true
+			}
+			wp.Libraries[i] = lib
+		}
+
+		pm.processes = append(pm.processes, wp)
+	}
+
+	// to make the crc check more meaningful
+	sort.Sort(pm.processes)
+	sort.Sort(pm.libraries)
+
+	return &pm
+}
+
+func NewProcessWatcher(match string, callback ChangeHandler) Watcher {
+	watcher := &processWatcher{
+		match:     match,
+		OnChange:  callback,
+		UpdatedAt: time.Now(),
+		pollSleep: env.PollSleep,
+	}
+
+	// Don't scan from here, we just end up with two running at once
+	return watcher
+}
+
+func NewAllProcessWatcher(callback ChangeHandler) Watcher {
+	return NewProcessWatcher("", callback)
+}
+
+func (pw *processWatcher) MarshalJSON() ([]byte, error) {
+	pw.Lock()
+	defer pw.Unlock()
+	return json.Marshal(map[string]interface{}{
+		"match":         pw.Match(),
+		"updated-at":    pw.UpdatedAt,
+		"being-watched": pw.BeingWatched,
+	})
+}
+
+func (wt *processWatcher) Start() {
+	wt.Lock()
+	wt.keepPolling = true
+	wt.Unlock()
+	go wt.listen()
+}
+
+func (wt *processWatcher) Stop() {
+	wt.Lock()
+	wt.keepPolling = false
+	wt.Unlock()
+}
+
+func (wt *processWatcher) Match() string {
```

Comment:
I feel this could be more descriptive somehow, esp with how it's reflected in the configuration.

```Go
	Command string `toml:"process"`
	Process string `toml:"match"`
```

This is already pretty "lol", since we have to backwards compat it. But how do you feel about (just throwing it out there), `inspect_process`?